odoo/fields.py

```diff
@@ -273,47 +277,77 @@ def __repr__(self):
+    # may not be shared.  In other words, those fields are never used directly,
+    # and are always recreated as toplevel fields.  On those fields, the base
+    # setup is useless, because only field.args is used for setting up other
+    # fields.  We therefore skip the base setup for those fields.  The only
+    # attributes of those fields are: '_sequence', 'args', 'model_name', 'name'
+    # and '_module', which makes their __dict__'s size minimal.
+
+    def __set_name__(self, owner, name):
+        """ Perform the base setup of a field. """
+        assert issubclass(owner, BaseModel)
+        self.model_name = owner._name
+        self.name = name
+        if getattr(owner, 'pool', None) is None:
+            # only for fields on classes defining models, not registry classes
+            self._module = owner._module
+            owner._base_fields.append(self)
+
+        if not self.args.get('related'):
+            self._direct = True
+        if self._direct or self._toplevel:
+            self._setup_attrs(owner, name)
+            if self._toplevel:
+                # free memory, self.args and self._base_fields are no longer useful
+                self.__dict__.pop('args', None)
+                self.__dict__.pop('_base_fields', None)
 
     #
     # Setup field parameter attributes
     #
 
-    def _can_setup_from(self, field):
-        """ Return whether ``self`` can retrieve parameters from ``field``. """
-        return isinstance(field, type(self))
-
-    def _get_attrs(self, model, name):
+    def _get_attrs(self, owner, name):
         """ Return the field parameter attributes as a dictionary. """
         # determine all inherited field attributes
-        modules = set()
         attrs = {}
-        if self.args.get('automatic') and resolve_mro(model, name, self._can_setup_from):
-            # prevent an automatic field from overriding a real field
-            self.args.clear()
-        if not (self.args.get('automatic') or self.args.get('manual')):
-            # magic and custom fields do not inherit from parent classes
-            for field in reversed(resolve_mro(model, name, self._can_setup_from)):
-                attrs.update(field.args)
-                if '_module' in field.args:
-                    modules.add(field.args['_module'])
-        attrs.update(self.args)         # necessary in case self is not in class
+        modules = []
+        for field in self.args.get('_base_fields', ()):
+            if not isinstance(field, type(self)):
+                attrs.clear()
+                modules.clear()
+                continue
+            attrs.update(field.args)
+            if field._module:
+                modules.append(field._module)
+        attrs.update(self.args)
+        if self._module:
+            modules.append(self._module)
 
         attrs['args'] = self.args
-        attrs['model_name'] = model._name
+        attrs['model_name'] = owner._name
```

Comment:
I wouldn't rename this parameter: 
- inconsistency with other `_get_attrs` overrides
- what kind of object is owner?
- Useless blame

Anyway if wonder is this parametter is really usefull since it isn't use in any override anymore. It could also be done in _setup_attrs. (name parametter could also be avoided without the dirty state hack)

Also, is this really usefull since model_name and name are set in `__set_name__`: https://github.com/odoo/odoo/pull/69924/files#diff-eb000d15355b38936d1b7f7506465ad081a3d96424c1f64dd1008ef3d4ac6427R309

Reply:
I renamed this parameter *on purpose*, since its type is different from the former one:
* before: `model` is an instance of the registry class of the given model
* after: `owner` is a model class (definition class or registry class) that holds the given field

> Also, is this really useful since model_name and name are set in __set_name__.
In theory, it shouldn't be useful.  In practice, it is.  For instance https://github.com/odoo/odoo/blob/14.0/addons/mail/wizard/mail_blacklist_remove.py#L10

Reply:
Ok then there are just some _get_attrs override that need to be adapted the same way in this case. (reminder)